pvlib/test/test_psm3.py

```diff
@@ -0,0 +1,57 @@
+"""
+test iotools for PSM3
+"""
+
+import os
+from pvlib.iotools import psm3
+from conftest import needs_pandas_0_22
+import numpy as np
+import pandas as pd
+import pytest
+from requests import HTTPError
+
+BASEDIR = os.path.abspath(os.path.dirname(__file__))
+PROJDIR = os.path.dirname(BASEDIR)
+DATADIR = os.path.join(PROJDIR, 'data')
+TEST_DATA = os.path.join(DATADIR, 'test_psm3.csv')
+LATITUDE, LONGITUDE = 40.5137, -108.5449
+HEADER_FIELDS = [
+    'Source', 'Location ID', 'City', 'State', 'Country', 'Latitude',
+    'Longitude', 'Time Zone', 'Elevation', 'Local Time Zone',
+    'Dew Point Units', 'DHI Units', 'DNI Units', 'GHI Units',
+    'Temperature Units', 'Pressure Units', 'Wind Direction Units',
+    'Wind Speed', 'Surface Albedo Units', 'Version']
+PVLIB_EMAIL = 'pvlib-admin@googlegroups.com'
+
+
+@needs_pandas_0_22
+def test_get_psm3():
+    """test get_psm3"""
+    header, data = psm3.get_psm3(LATITUDE, LONGITUDE, 'DEMO_KEY', PVLIB_EMAIL)
+    expected = pd.read_csv(TEST_DATA)
+    # check datevec columns
+    assert np.allclose(data.Year, expected.Year)
+    assert np.allclose(data.Month, expected.Month)
+    assert np.allclose(data.Day, expected.Day)
+    assert np.allclose(data.Hour, expected.Hour)
+    assert np.allclose(data.Minute, expected.Minute)
+    # check data columns
+    assert np.allclose(data.GHI, expected.GHI)
+    assert np.allclose(data.DNI, expected.DNI)
+    assert np.allclose(data.DHI, expected.DHI)
+    assert np.allclose(data.Temperature, expected.Temperature)
+    assert np.allclose(data.Pressure, expected.Pressure)
+    assert np.allclose(data['Dew Point'], expected['Dew Point'])
+    assert np.allclose(data['Surface Albedo'], expected['Surface Albedo'])
+    assert np.allclose(data['Wind Speed'], expected['Wind Speed'])
+    assert np.allclose(data['Wind Direction'], expected['Wind Direction'])
+    # check header
+    for hf in HEADER_FIELDS:
+        assert hf in header
+    # check timezone
+    assert (data.index.tzinfo.zone == 'Etc/GMT%+d' % -header['Time Zone'])
+    # check errors
+    with pytest.raises(HTTPError):
+        psm3.get_psm3(LATITUDE, LONGITUDE, api_key='BAD', email=PVLIB_EMAIL)
+    with pytest.raises(HTTPError):
```

Comment:
why does this one fail?

Reply:
* `api_key`="BAD" fails because the response is HTTP/1.1 403 Forbidden and there's a long error message in the content, no JSON, that says something like, "please go to NREL developer network and register for a key"
* Bristol-UK fails because PSM3 only works in the US, maybe South America, and I think some parts of India perhaps around Chennai? I can confirm - perhaps this should be a warning in the docstring?

perhaps I should add some more comments here to explain these tests - like I did here - for future devs?

Reply:
BAD was self explanatory. I don't think we need to test that it fails outside the US. Not our problem.

Reply:
PSM3 has data in a lat/long rectangle around the lower 48 states, so there's coverage in northern Mexico for example.  But it has a coastline filter and excludes data in the Gulf of Mexico.  I think we should test for failure at lat/long where PSM3 does not have data.

Reply:
Does PSM3 return a sensible error if lat/lon outside of its area?

Reply:
```
from get_psm3 import get_psm3

import api_keys

MY_KEY = api_keys.get_api_key('nsrdb')

for lat, lon in zip([25, 25, 25], [-125, -124.5, -124]):
    try:
        meta, data = get_psm3(lat, lon, names='tmy', interval=60,
                              api_key=MY_KEY)
    except Exception as e:
        print("{}, {}: ".format(lat, lon), e)
```

Returns
```
25, -125:  ["The 'wkt' parameter did not match any known data points."]
25, -124.5:  Expecting value: line 1 column 1 (char 0)
25, -124:  Expecting value: line 1 column 1 (char 0)
```

The first line is what I expect, the next 2 are json decoding errors that puzzle me.

Reply:
Aha, I suspect the 2nd and 3rd messages are empty responses because of the once-per-2second  [throttle](https://developer.nrel.gov/docs/solar/nsrdb/guide/), so `raise requests.HTTPError(response.json()['errors'])` is throwing the exception about Expecting value:...

Reply:
> ["The 'wkt' parameter did not match any known data points."]

This could be confusing to pvlib users because we're asking the to specify lat/lon instead of API. So I'm ok with catching this in the function and raising a more descriptive error about lat/lon.